proposals/readonly-struct-init.md

```diff
@@ -0,0 +1,161 @@
+
+This feature will allow for assigning of `readonly` fields and get only auto-implemented properties
+in object initialezrs when the type being constructed is a `struct` or `ref struct`.
+
+## Motivation
+
+The intent of `readonly` in C# is to prevent the modification of a value after construction has 
+completed. That is once the code calling `new` has a reference to the value there should be no 
+further modification of `readonly` fields. Until the calling code has a reference to the object
+modifications can continue to be made via `this`.
+
+This makes it even easier to define usable `readonly struct` types in C#. The simplest form of
+declaring the fields now creates a fully usable value:
+
+``` csharp
+readonly struct Point { 
+    readonly int X;
+    readonly int Y; 
+}
+
+var p = new Point { X = 42; Y = 13 };
+```
+
+This means authoring `readonly struct` types in C# is just as easy as creating the less
+perferred mutable `struct` types. There is no longer a need to define a constructor which simply 
+re-lists every field as a parameter and then assigns them one at a time. The definition of the 
+field itself allows provides a mechanism for initialization
+
+## Detailed Design 
+
+### Language
+
+The language will extend the object initialization rules to allow for an accessible readonly field 
+or auto-implemented property to appear in the member initialization list. In such a context the 
+members will allow for assignment just as it would inside a constructor of the object. 
+
+The following restrictions will apply to this extension:
+
+1. The type being instantiated must be a `struct` or `ref struct`. 
+2. The constructor used in the initializer cannot have any `out` or `ref` parameters.
+
+The (2) restriction exists because such a parameter could smuggle out `this` from the object 
+constructor. That would make a reference to the constructed object visible to the code calling
+`new` and hence allow it to observe mutations to the value.
+
+### Compiler 
+
+The compiler will change how the backing field of an auto-implemented property is emitted in the 
+following ways:
+
+- The accessibility of the backing field will match the accessibility of the property.
+- The name of the field will be unspeakable (illegal C# identifier hence never imported) but it 
+will be possible to map such fields to the properties that defined them.
+
+Taken together this allows the C# compiler to translate a get only auto-implemented property into
+a simple field initialization when used as an object initializer. 
+
+``` csharp 
+struct Name {
+    // Backing field: public readonly string <FirstName>__BackingField;
+    public string FirstName { get; }
+    // Backing field: public readonly string <LastName>__BackingField;
+    public string LastNmae { get; }
+}
+
+var name = new Name { FirstName = "Jared", LastName = "Parsons" };
+
+// Emitted as 
+var name = new Name { <FirstName>__BackingField = "Jared", <LastName>__BackingField = "Parsons" };
```

Comment:
We're relying on the runtime not blocking assignment after construction?

Reply:
This violates the CLI specification. A conforming implementation could reject the code as invalid. The spec just says that the runtime "need not" enforce this on unverifiable code, but not that it cannot enforce it. This is different from other cases, like overlapping object references, which are called out as unverifiable but valid. 

II.16.1.2
> In unverifiable code, the VES need not check whether initonly fields are mutated outside the constructors. The VES need not report any errors if a method changes the value of a constant. However, such code is not valid. 

I would also say this goes beyond other cases where I've supported deviating from the spec, like where it does not allow custom modifiers on type arguments. Those cases are spec bugs or omissions IMHO, where the consequences were likely unintended. Here, however, it is very clear that such code was never intended to be deemed valid.

Reply:
> We're relying on the runtime not blocking assignment after construction?

No. The fields are still marked as `readonly` hence they are blocked by default. Only compilers which understood the new verification rules would be able to participate in this. In theory at least. 

Note: this is mostly meant as an exercise of "if we wanted to verify this could we"? It's hard to be 100% certain of this until we have a working verifier. But for a number of features (ref locals in particular) we sketched out what we thought the rules would be to make sure a) we weren't ignoring this space and b) felt confident rules could be written on top of our feature that would be verifiable.

Reply:
>> We're relying on the runtime not blocking assignment after construction?

> No. The fields are still marked as readonly hence they are blocked by default

I think David meant that the codegen for the initialization would be relying on the fact that the runtime doesn't enforce initonly rules (unless the verifier is required to run).

And that is relying on an implementation detail of the runtime since the spec says that code that does this is "not valid".  It also says that the runtime "need not" enforce it. but if this change goes in, then the runtime "must not" enforce it.

I think this is worth calling out: we wouldn't just be relying on code that is currently unverifiable, we'd be relying on code that is currently **invalid**.